### ISSUE TITLE
How is the Hostname of the root service in TrafficSplit going to be resolvable?

### ISSUE THREAD

**Issue by shashankram** [servicemeshinterface/smi-spec#160]
Consider the following example.
```
apiVersion: split.smi-spec.io/v1alpha3
kind: TrafficSplit
metadata:
  name: canary
spec:
  service: website.com
  backends:
  - service: website-v1
    weight: 90
  - service: website-v2
    weight: 10
```

When a service in K8s curls `http://website.com`, since `website.com` is just a hostname for a root service that K8s is not aware of, how would this be resolvable? Are users expected to configure DNS for the domain `website.com` here?

**Comment by grampelberg** [comment]
The `service` must be a service in the same namespace. `website.com` is not valid as it is not a k8s resource.

**Comment by shashankram** [comment]
> The `service` must be a service in the same namespace. `website.com` is not valid as it is not a k8s resource.

Thanks for clarifying, I was wondering if the spec allows for creating aliases to service.

**Comment by grampelberg** [comment]
Would you mind explaining your use case a little bit more? Doing a traffic split from `website.com` to two internal services doesn't seem like something anyone would want to do?

**Comment by shashankram** [comment]
> Would you mind explaining your use case a little bit more? Doing a traffic split from `website.com` to two internal services doesn't seem like something anyone would want to do?

I've read about a few use cases where users of Istio have tried to use the `hosts` in `VirtualService` as an alias for actual K8s services, so was wondering how it fits here.

I am still a bit confused, consider the following example where `website` is a k8s service running in the namespace `website-ns`. Given the following traffic split, how would a service `foo` running in the namespace `foo-ns` do a traffic split for `website`? For `website` to be resolvable via DNS, the FQDN must be such that the queries from `foo` to `website` are resolvable. Would this mean the service `foo` be aware of the namespace `website` belongs to for the domain in the HTTP request to resolve to a valid IP?

**Comment by grampelberg** [comment]
I'm still super confused. `website` has an FQDN of `website.website-ns.svc.cluster.local`. That is resolvable via. DNS because it is exported by default either with `kube-dns` or `coredns`.

If you're talking about public DNS, let's say `website.com`, you would write an `Ingress` resource. This would contain the mapping from `website.com` to `website`. The ingress controller would then look up the endpoints to `website` in the `website-ns` namespace and forward the request there. Notice that no DNS lookups actually happened inside the ingress controller. If the ingress controller implemented TrafficSplit, it would then lookup the leaf services, fetch the endpoints for those and do load balancing. If a service mesh was on the *outbound* side of the ingress controller, it would do something basically the same.

**Comment by shashankram** [comment]
I am not referring to an external endpoint, but in-cluster. Was just wondering whether there is a way for apps to not care about the namespace of destination services, `website` in this example by using an alias. If the user decides to change the namespace for the `website` service, a service `foo` making HTTP requests to `website` would need to be aware of its namespace to use an FQDN that is resolvable as `website.website-ns.svc.cluster.local`.

**Comment by grampelberg** [comment]
Yes, that is a limitation of k8s.

**Comment by shashankram** [comment]
> Yes, that is a limitation of k8s.

So much for decoupling app logic from networking :-)

**Comment by shashankram** [comment]
@grampelberg , one clarification - do the backends in a split have to be a k8s service or can they just be deployments with selector `app=<root-service>` ?

**Comment by grampelberg** [comment]
Backends must be service resources as well.

**Comment by shashankram** [comment]
> Backends must be service resources as well.

In that case, even though a client of the `website` service is making requests to `http://website.website-ns.svc.cluster.local`, when a split with a different backend service is applied, for ex `website-v2` as the backend for apex `website`, the client would still continue using the root service FQDN in HTTP requests but the service mesh implementation should use the `website-v2` backend to redirect traffic. 
In other words, the clients always use the FQDN of the root service, and the service mesh implementation needs to split traffic to the correct backend. So the IP in the request packet originating from the client might not correspond to the actual backend defined by the Split, but just used to resolve to a service IP, after which the service mesh implementation does the magic. Is this correct?

**Comment by grampelberg** [comment]
That is correct.

**Comment by shashankram** [comment]
We should document this for clarity. Thanks @grampelberg for clarifying.

**Comment by shashankram** [comment]
@grampelberg , I think `spec.service` should just be an FQDN that is resolvable within the cluster.

**Comment by grampelberg** [comment]
@shashankram I am very strongly against that. At the bare minimum it would need to be named something different (fqdn maybe). How do the endpoints get updated? Where does the implementation fetch data from?

**Comment by shashankram** [comment]
I don't think the service mesh implementation should worry about the FQDN at all. The service mesh implementation would fetch the endpoints for a top level service from its backends. Making the top level service an FQDN would be clear and avoid weird use cases where a leaf and top level service are the same, which could imply recursion.

We should rename `spec.service` to `spec.host`, and let the users worry about how that is resolvable - whether they want to create a top level service just for resolving to an IP the service mesh implementation doesn't use, or bring their own DNS for it.

**Comment by shashankram** [comment]
@grampelberg could you reopen this issue while this is still in discussion. I can't seem to reopen it.

**Comment by grampelberg** [comment]
> I don't think the service mesh implementation should worry about the FQDN at all. 

Yup, this is why I'm uninterested in doing this. If you'd like to make this proposal, please open a separate issue or PR. I'd like to know what problem you're solving as I have a long list of reasons why `spec.host` is a bad idea in absence of problems being solve.

**Comment by shashankram** [comment]
Consider the following split
```
apiVersion: split.smi-spec.io/v1alpha3
kind: TrafficSplit
metadata:
  name: canary
  namespace: website-ns
spec:
  service: website
  backends:
  - service: website-v1
    weight: 90
  - service: website-v2
    weight: 10
```

The assumption here is that `website` is a k8s service. The problem with this assumption is that clients have to hardcode a service name, that might no longer be even relevant. In this example, `spec.service` could be anything, it could even be `website-1`, and `website-1` could be versioned out with clients still referencing it, which is misleading. Moreover clients in different namespaces would need to access the `website` service using their FQDN `website.website-ns`. So just referencing the service name for `spec.service` is not enough. Clearly `spec.service` needs to be FQDN here.

**Comment by grampelberg** [comment]
Clients have to hardcode FQDNs though, what's the benefit?

**Comment by shashankram** [comment]
> Clients have to hardcode FQDNs though, what's the benefit?
> 
> In your example, you'd just create a new traffic split (which you would have to do whether you're using FQDNs or not).
> 
> The downside to using FQDNs is that they must match. Some questions that come from that:
> 
> * What happens if someone's using a non-standard DNS name for their cluster (`foo.local`)?
> * How do you choose between someone's client saying `website` vs `website.default` vs `website.default.svc.foo.local`? All three would need new traffic splits to support each, and would fail in transparent ways to the users.

This would be true even if the clients are in a different namespace than `website`. 
- A client in `foo` namespace  trying to access `website` in `website-ns` namespace would need to access it using FQDN `website.website-ns`. 
- A client in `website-ns` namespace would access website using FQDN `website`
- A client in a different cluster would need to access website using `website.website-ns.svc.<cluster>`

So depending on the location of the client, different splits would be required. If `spec.service` is FQDN of the form `service.namespace.cluster.domain`, all clients can use the same FQDN.

> * How does the implementation discover what it should be load balancing?

The implementation would do load balancing based on the backends specified for the top level FQDN. The backend are services with endpoints.

> * How does validation happen?

No validation is required for the FQDN. It should be the user's responsibility to ensure the FQDN is resolvable by the client. One important thing to note here is that the FQDN would resolve to some IP, doesn't matter what, its the backends for the FQDN that the traffic would be routed too. This
is still the case with `spec.service` being a service, as noted in our discussion above.

> * How is policy implemented? Would it be possible to have an attacker send traffic transparently?
> 

Could you explain the scenario more in detail?

> I'll reiterate, I'd really love to know what the problem is that you're trying to solve. I suspect that there's a creative way we can get a solution to that without opening pandora's box for `TrafficSplit` implementations and use cases.

**Comment by grampelberg** [comment]
I'm super confused. Maybe we could get on a call some time and chat through it? Might be a great topic to bring up during the next community call!

**Comment by shashankram** [comment]
Sorry @grampelberg for not clearing this up.
Let's take one question at a time. What are your confusions?

**Comment by grampelberg** [comment]
It just feels like there's some misunderstandings that'll require a higher bandwidth conversation is all. I basically have a ton of questions and so far I'm just getting more confused =( I'm in [slack](https://cloud-native.slack.com/messages/smi) if you're interested, though next week would probably work better for me.

**Comment by shashankram** [comment]
Let's chat over slack next week!
